src/hotspot/share/opto/superword.cpp

```diff
@@ -1384,60 +1402,198 @@ void SuperWord::combine_pairs_to_longer_packs() {
+      Node_List* second_pack = status.second_pack();
+      _packset.at_put(i, nullptr); // take out pack
+      if (first_pack != nullptr) {
+        // The first pack can be put at the current position
+        assert(i >= new_packset_length, "only move packs down");
+        _packset.at_put(new_packset_length++, first_pack);
+      }
+      if (second_pack != nullptr) {
+        // The second node has to be appended at the end
+        _packset.append(second_pack);
+      }
+    }
+    _packset.trunc_to(new_packset_length);
+  } while (changed);
 
 #ifndef PRODUCT
   if (is_trace_superword_packset()) {
-    tty->print_cr("\nAfter Superword::split_packs_longer_than_max_vector_size");
+    tty->print_cr("\nAfter %s", split_name);
     print_packset();
   }
 #endif
 }
 
+// Split packs at boundaries where left and right have different use or def packs.
+void SuperWord::split_packs_at_use_def_boundaries() {
+  split_packs("SuperWord::split_packs_at_use_def_boundaries",
+               [&](const Node_List* pack) {
+                 uint pack_size = pack->size();
+                 uint boundary = find_use_def_boundary(pack);
+                 assert(boundary < pack_size, "valid boundary %d", boundary);
+                 if (boundary != 0) {
+                   return SplitTask::make_split(pack_size - boundary, "found a use/def boundary");
+                 }
+                 return SplitTask::make_unchanged();
+               });
+}
+
+// Split packs that are only implemented with a smaller pack size. Also splits packs
+// such that they eventually have power of 2 size.
+void SuperWord::split_packs_only_implemented_with_smaller_size() {
+  split_packs("SuperWord::split_packs_only_implemented_with_smaller_size",
+               [&](const Node_List* pack) {
+                 uint pack_size = pack->size();
+                 uint implemented_size = max_implemented_size(pack);
+                 if (implemented_size == 0)  {
+                   return SplitTask::make_rejected("not implemented at any smaller size");
+                 }
+                 assert(is_power_of_2(implemented_size), "power of 2 size or zero: %d", implemented_size);
+                 if (implemented_size != pack_size) {
+                   return SplitTask::make_split(implemented_size, "only implemented at smaller size");
+                 }
+                 return SplitTask::make_unchanged();
+               });
+}
+
+// Split packs that have mutual dependence, until all packs are mutually_independent.
+void SuperWord::split_packs_to_break_mutual_dependence() {
+  split_packs("SuperWord::split_packs_to_break_mutual_dependence",
+               [&](const Node_List* pack) {
+                 uint pack_size = pack->size();
+                 assert(is_power_of_2(pack_size), "ensured by earlier splits %d", pack_size);
+                 if (!is_marked_reduction(pack->at(0)) &&
+                     !mutually_independent(pack)) {
+                   // Split in half.
```

Comment:
Maybe you could add a comment here that splitting in half is a best guess/intuitive way to continue